Scope variables: static/extern mismatch between .cpp and header

## Description

When generating C++ output with `--cpp` flag for a scope that has private variables, the transpiler generates:

- **Header (.h)**: `extern uint32_t crc32_offset;`
- **Implementation (.cpp)**: `static uint32_t crc32_offset = 0;`

This causes a compilation warning and is semantically inconsistent:

```
src/crc32.cpp:23:17: warning: 'crc32_offset' was declared 'extern' and later 'static' [-fpermissive]
   23 | static uint32_t crc32_offset = 0;
      |                 ^~~~~~~~~~~~
src/crc32.h:20:17: note: previous declaration of 'crc32_offset'
   20 | extern uint32_t crc32_offset;
      |                 ^~~~~~~~~~~~
```

## Source Code

```cnx
scope crc32 {
    u32 offset <- 0;
    u8 length <- 0;

    public u32 calculateChecksum(const AppConfig config) {
        // ...
    }
}
```

## Expected Behavior

The header and implementation should be consistent. Options:
1. Both `extern` in header, non-static in .cpp (allows external access)
2. Don't expose private scope variables in the header at all (if they're truly private)

## Workaround

Manually edit the .cpp to remove `static` keyword from scope variables.

## Environment

- cnext version: 0.1.8
- Platform: Linux
- Compiler: arm-none-eabi-g++ 11.3.1 (Teensy/PlatformIO)

--- a/src/symbols/CNextSymbolCollector.ts
+++ b/src/symbols/CNextSymbolCollector.ts
@@ -77,27 +77,38 @@ class CNextSymbolCollector {
 
     // Collect scope members
     for (const member of scopeDecl.scopeMember()) {
+      // Issue #218: Check visibility modifier - default is "private"
+      const visibilityMod = member.visibilityModifier();
+      const visibility = visibilityMod?.getText() ?? "private";
+      const isPublic = visibility === "public";
+
       if (member.variableDeclaration()) {
-        this.collectVariable(member.variableDeclaration()!, name);
+        this.collectVariable(member.variableDeclaration()!, name, isPublic);
       }
       if (member.functionDeclaration()) {
-        this.collectFunction(member.functionDeclaration()!, name);
+        this.collectFunction(member.functionDeclaration()!, name, isPublic);
       }
       // ADR-034: Handle bitmap declarations in scopes
       if (member.bitmapDeclaration()) {
-        this.collectBitmap(member.bitmapDeclaration()!, name);
+        this.collectBitmap(member.bitmapDeclaration()!, name, isPublic);
       }
       // Handle register declarations in scopes
       if (member.registerDeclaration()) {
-        this.collectScopedRegister(member.registerDeclaration()!, name);
+        this.collectScopedRegister(
+          member.registerDeclaration()!,
+          name,
+          isPublic,
+        );
       }
     }
   }
 
   // Collect scoped register declarations (e.g., scope.register -> Scope_Register)
+  // Issue #218: isPublic parameter controls header export
   private collectScopedRegister(
     reg: Parser.RegisterDeclarationContext,
     scopeName: string,
+    isPublic: boolean = true,
   ): void {
     const regName = reg.IDENTIFIER().getText();
     const fullName = `${scopeName}_${regName}`;
@@ -109,7 +120,7 @@ class CNextSymbolCollector {
       sourceFile: this.sourceFile,
       sourceLine: line,
       sourceLanguage: ESourceLanguage.CNext,
-      isExported: true,
+      isExported: isPublic,
       parent: scopeName,
     });
 
@@ -127,7 +138,7 @@ class CNextSymbolCollector {
         sourceFile: this.sourceFile,
         sourceLine: memberLine,
         sourceLanguage: ESourceLanguage.CNext,
-        isExported: true,
+        isExported: isPublic,
         parent: fullName,
         accessModifier: accessMod,
       });
@@ -183,9 +194,11 @@ class CNextSymbolCollector {
   }
 
   // ADR-034: Collect bitmap declarations
+  // Issue #218: isPublic parameter controls header export
   private collectBitmap(
     bitmap: Parser.BitmapDeclarationContext,
     parent: string | undefined,
+    isPublic: boolean = true,
   ): void {
     const name = bitmap.IDENTIFIER().getText();
     const line = bitmap.start?.line ?? 0;
@@ -200,7 +213,7 @@ class CNextSymbolCollector {
       sourceFile: this.sourceFile,
       sourceLine: line,
       sourceLanguage: ESourceLanguage.CNext,
-      isExported: true,
+      isExported: isPublic,
       parent,
     });
 
@@ -230,7 +243,7 @@ class CNextSymbolCollector {
         sourceFile: this.sourceFile,
         sourceLine: fieldLine,
         sourceLanguage: ESourceLanguage.CNext,
-        isExported: true,
+        isExported: isPublic,
         parent: fullName,
         signature: `${bitRange} (${width} bit${width > 1 ? "s" : ""})`,
       });
@@ -239,9 +252,11 @@ class CNextSymbolCollector {
     }
   }
 
+  // Issue #218: isPublic parameter controls header export
   private collectFunction(
     func: Parser.FunctionDeclarationContext,
     parent: string | undefined,
+    isPublic: boolean = true,
   ): void {
     const name = func.IDENTIFIER().getText();
     const line = func.start?.line ?? 0;
@@ -278,7 +293,7 @@ class CNextSymbolCollector {
       sourceFile: this.sourceFile,
       sourceLine: line,
       sourceLanguage: ESourceLanguage.CNext,
-      isExported: true,
+      isExported: isPublic,
       parent,
       signature,
       parameters,
@@ -306,9 +321,11 @@ class CNextSymbolCollector {
     }
   }
 
+  // Issue #218: isPublic parameter controls header export
   private collectVariable(
     varDecl: Parser.VariableDeclarationContext,
     parent: string | undefined,
+    isPublic: boolean = true,
   ): void {
     const name = varDecl.IDENTIFIER().getText();
     const line = varDecl.start?.line ?? 0;
@@ -335,7 +352,7 @@ class CNextSymbolCollector {
       sourceFile: this.sourceFile,
       sourceLine: line,
       sourceLanguage: ESourceLanguage.CNext,
-      isExported: true,
+      isExported: isPublic,
       parent,
       size,
     });

--- a/tests/integration/scope-header-visibility.test.ts
+++ b/tests/integration/scope-header-visibility.test.ts
@@ -1,0 +1,174 @@
+#!/usr/bin/env tsx
+/**
+ * Integration test for Issue #218: Scope variables static/extern mismatch
+ * Tests that private scope variables do NOT appear in generated headers
+ * while public scope variables DO appear with extern declarations.
+ */
+
+import { readFileSync, existsSync, writeFileSync, mkdirSync } from "fs";
+import { join, dirname } from "path";
+import { fileURLToPath } from "url";
+import Pipeline from "../../src/pipeline/Pipeline";
+
+const __filename = fileURLToPath(import.meta.url);
+const __dirname = dirname(__filename);
+
+// Create a temporary test file content
+const testSource = `
+// Test: Issue #218 - Private scope variables should not appear in headers
+scope TestScope {
+    // Private variables (default - no modifier)
+    u32 privateCounter <- 0;
+    bool privateFlag <- false;
+
+    // Public variables (explicit public modifier)
+    public u32 publicCounter <- 10;
+    public bool publicFlag <- true;
+
+    // Private method
+    u32 getPrivateInternal() {
+        return this.privateCounter;
+    }
+
+    // Public method accessing private state
+    public u32 getPrivateCounter() {
+        return this.privateCounter;
+    }
+
+    // Public method accessing public state
+    public u32 getPublicCounter() {
+        return this.publicCounter;
+    }
+}
+
+u32 main() {
+    return 0;
+}
+`;
+
+// Main test function
+async function runTest() {
+  // Write test source to temp location
+  const testDir = "/tmp/c-next-test-218";
+  const testFile = join(testDir, "scope-visibility.cnx");
+
+  mkdirSync(testDir, { recursive: true });
+  writeFileSync(testFile, testSource, "utf-8");
+
+  // Configure pipeline with C++ output and header generation
+  const pipeline = new Pipeline({
+    inputs: [testFile],
+    outDir: testDir,
+    includeDirs: [],
+    generateHeaders: true,
+    cppRequired: true, // Force C++ output
+  });
+
+  console.log("Compiling with C++ output and header generation...");
+  const result = await pipeline.run();
+
+  if (!result.success) {
+    console.error("Compilation failed:");
+    console.error("Errors:", result.errors);
+    process.exit(1);
+  }
+
+  // Read the generated header file
+  const headerFile = join(testDir, "scope-visibility.h");
+  if (!existsSync(headerFile)) {
+    console.error("ERROR: Header file was not generated");
+    process.exit(1);
+  }
+
+  const headerContent = readFileSync(headerFile, "utf-8");
+
+  console.log("\n=== Generated Header Content ===");
+  console.log(headerContent);
+  console.log("================================\n");
+
+  // Also read the generated C++ file to see the implementation
+  const cppFile = join(testDir, "scope-visibility.cpp");
+  if (existsSync(cppFile)) {
+    const cppContent = readFileSync(cppFile, "utf-8");
+    console.log("\n=== Generated C++ Content ===");
+    console.log(cppContent);
+    console.log("==============================\n");
+  }
+
+  // Validation checks
+  const checks = [
+    {
+      test: () =>
+        headerContent.includes("extern uint32_t TestScope_publicCounter"),
+      description:
+        "Public variable TestScope_publicCounter should have extern declaration",
+      expected: true,
+    },
+    {
+      test: () => headerContent.includes("extern bool TestScope_publicFlag"),
+      description:
+        "Public variable TestScope_publicFlag should have extern declaration",
+      expected: true,
+    },
+    {
+      test: () => !headerContent.includes("TestScope_privateCounter"),
+      description:
+        "Private variable TestScope_privateCounter should NOT appear in header",
+      expected: true,
+    },
+    {
+      test: () => !headerContent.includes("TestScope_privateFlag"),
+      description:
+        "Private variable TestScope_privateFlag should NOT appear in header",
+      expected: true,
+    },
+    {
+      test: () => headerContent.includes("TestScope_getPrivateCounter"),
+      description:
+        "Public method TestScope_getPrivateCounter should have prototype",
+      expected: true,
+    },
+    {
+      test: () => headerContent.includes("TestScope_getPublicCounter"),
+      description:
+        "Public method TestScope_getPublicCounter should have prototype",
+      expected: true,
+    },
+    {
+      test: () => !headerContent.includes("TestScope_getPrivateInternal"),
+      description:
+        "Private method TestScope_getPrivateInternal should NOT appear in header",
+      expected: true,
+    },
+  ];
+
+  let passed = 0;
+  let failed = 0;
+
+  for (const check of checks) {
+    const checkResult = check.test();
+    if (checkResult === check.expected) {
+      console.log(`PASS: ${check.description}`);
+      passed++;
+    } else {
+      console.error(`FAIL: ${check.description}`);
+      failed++;
+    }
+  }
+
+  console.log(`\nResults: ${passed} passed, ${failed} failed`);
+
+  if (failed > 0) {
+    console.error(
+      "\nTest FAILED - private scope members are incorrectly appearing in the header",
+    );
+    process.exit(1);
+  }
+
+  console.log("\nAll checks passed!");
+}
+
+runTest().catch((err) => {
+  console.error("Test error:", err);
+  process.exit(1);
+});

